nixos/modules/services/networking/headscale.nix

```diff
@@ -0,0 +1,486 @@
+{ config, lib, pkgs, ... }:
+with lib;
+let
+  cfg = config.services.headscale;
+
+  mkNotNull = value:
+    optional (value != null) value;
+
+  settings = {
+    server_url = cfg.serverUrl;
+    listen_addr = "${cfg.address}:${toString cfg.port}";
+
+    private_key_path = cfg.privateKeyFile;
+
+    derp = {
+      urls = cfg.derp.urls;
+      paths = cfg.derp.paths;
+      auto_update_enable = cfg.derp.autoUpdate;
+      update_frequency = cfg.derp.updateFrequency;
+    };
+
+    # Turn off update checks since the origin of our package
+    # is nixpkgs and not Github.
+    disable_check_updates = true;
+
+    ephemeral_node_inactivity_timeout = cfg.ephemeralNodeInactivityTimeout;
+
+    db_type = cfg.database.type;
+    db_path = cfg.database.path;
+    db_host = mkNotNull cfg.database.host;
+    db_port = mkNotNull cfg.database.port;
+    db_name = mkNotNull cfg.database.name;
+    db_user = mkNotNull cfg.database.user;
+
+    log_level = cfg.logLevel;
+
+    dns_config = {
+      nameservers = cfg.dns.nameservers;
+      domains = cfg.dns.domains;
+      magic_dns = cfg.dns.magicDns;
+      base_domain = cfg.dns.baseDomain;
+    };
+
+    unix_socket = "${cfg.runDir}/headscale.sock";
+
+    # OpenID Connect
+    oidc = {
+      issuer = cfg.openIdConnect.issuer;
+      client_id = cfg.openIdConnect.clientId;
+      domain_map = cfg.openIdConnect.domainMap;
+    };
+
+    tls_letsencrypt_hostname = mkNotNull cfg.tls.letsencrypt.hostname;
+    tls_letsencrypt_cache_dir = "${cfg.dataDir}/.cache";
+    tls_letsencrypt_challenge_type = mkNotNull cfg.tls.letsencrypt.challengeType;
+    tls_letsencrypt_listen = mkNotNull cfg.tls.letsencrypt.listen;
+    tls_cert_path = mkNotNull cfg.tls.certFile;
+    tls_key_path = mkNotNull cfg.tls.keyFile;
+
+    acl_policy_path = cfg.aclPolicyFile;
+  };
+
+  settingsFormat = pkgs.formats.yaml { };
+  configFile = settingsFormat.generate "headscale.yaml" settings;
+in
+{
+  options = {
+    services.headscale = {
+      enable = mkEnableOption ''
+        Whether to enable the headscale coordination server for Tailscale.
+      '';
+
+      package = mkOption {
```

Comment:
since recently there is a `mkPackageOption` similar to `mkEnableOption` to make these options nicer to declare, if you like that :)